### DISCUSSION TITLE
VpnHood Site-To-Site

### DISCUSSION THREAD

**Discussion by asoMansoury** [vpnhood/VpnHood#285] [Ideas]
I feel for having a more impressive VPN server it is really critical to provide Site-To-Site VPN between VpnHood Servers.
Let me give you an example to know how much this feature is so important in countries like Iran and China.
For example, consider I have an Ubuntu server in Iran and install VpnHood Server in Iran. In such a situation when I give Tokens to my clients they will be able to connect to my Iran's VpnHood Server. By connecting my VpnHood server which is located in Iran, to another VpnHood in Foreign Country, then my clients will be able to break all the rules which have been defined by governors, especially when they block the whole international Internet. I don't know if you were aware that my country's Internet has been blocked  for two weeks! Not just one time, two times!!!. Of course, we were allowed to use our national network which is known as the national net.
By providing this feature your VPN Server Will be the most well-known VPN Server in Iran and China and their rules for blocking the Internet will be useless easily!

**Comment by trudyhood** [comment]
Hi,

I am wondering how your server in Iran has access to your VPN server, while your client doesn't have direct access to your VPN hood server. It seems like you have special privileges on some servers that your clients don't have, resulting in potential danger if you take advantage of your privilege and run such a server.

However, technically, you don't need to run a double VPN hood server and connect one to another like a bridge. You can achieve that by simply forwarding all ports from your server (located in your country) to the VPN hood server.

I am not an expert in configuring this, but there are many tools and documents available. In this scenario, your local server acts as a bridge, such as the solution suggested in this link: https://unix.stackexchange.com/questions/57721/how-to-forward-a-port-from-one-machine-to-another

Please feel free to share any other ideas you have.

**Comment by asoMansoury** [comment]
Hi, No I don't have any privilege for accessing the Internet.In Iran
meanwhile the International Internet is blocked, accessing for the internet
on data centers like of Asiatech is not restricted, and Iranian clients can
take a server on there personally and access to the Internet.
So by having this feature it would be possible to bypass filtering.

**Reply by trudyhood** [reply]
Good, I would be grateful if you share the result of the port forwarding I suggested.

**Comment by asoMansoury** [comment]
Hi sure, I have to provide one server in Iran , after that I would share
that with you

**Reply by trudyhood** [reply]
Ok, Just to forward TCP 443, if you are successful, we will try to change UDP and make VpnHood works on a single UDP port to be forward more easily.

**Reply by asoMan1369** [reply]
Do I have to Install VpnHood on Iran's server and then port forward the whole traffic to German Server which VpnHood is installed?
Then via Console Manager  create tokens on Iran's server for clients, am i right?

**Reply by trudyhood** [reply]
No, reverse.  You should  Install VpnHood server on your German server and forward the TCP port from the Iran server to the German server. In other words, Port forwarding must be installed on your Iran server to bridge the connection to the VpnHood server on the German server.

**Reply by asoMan1369** [reply]
I think this idea does not work properly, because when I'm going to create a token on Console Manager, then the token will contain the IP Address of German server, and I think the clients will connect to German not Iran, so the traffic would not go from Iran to German Server.

**Comment by trudyhood** [comment]
Yes, but you can configure the server manually and add your iran server ip as PublicInToken in server access-point manager. Set server group to manual to be able doing that. 

I'm with you if you encounter any issue from VpnHood server side.

**Reply by asoMan1369** [reply]
As I got from your guidance, at first I have to install Ubuntu Server on Iran, then port forwarding all the traffic to German server. 
After that, I have to install vpnhood server via VpnHood Access Server on My German server and give tokens to my clients which I should set Iran's IP on tokens manually. 
So by doing this scenario all traffic will go to German server through Iran. 
OK I will doing that, but I'm not sure it works, I will let you know whenever I set up all.

**Comment by asoMan1369** [comment]
If you have Telegram give me, your Telegram ID, to be in touch with you.

**Reply by trudyhood** [reply]
> If you have Telegram give me, your Telegram ID, to be in touch with you.

I am committed to working in an open community on this project, but if you are going to send us some credentials, you can send them to support@vpnhood.com

Although I am curious to see your job done, I believe it works and is possible. You need something like a pure TCP forwarder or PORT forwarder with NAT support without any encryption or handshake.

https://unix.stackexchange.com/questions/10428/simple-way-to-create-a-tunnel-from-one-local-port-to-another
I hope to have good news from you.

**Reply by asoMansoury** [reply]
I runned this command on my ubuntu server to Nat all requests which comes on 443 port of my Iran Server and reforward it to my Germand Server on 443 Port.

iptables -t nat -A PREROUTING -p tcp --dport 443 -j DNAT --to-destination GermanServer:443

**Reply by asoMansoury** [reply]
I used below link as my reference.
https://serverfault.com/questions/586486/how-to-do-the-port-forwarding-from-one-ip-to-another-ip-in-same-network
I didn't add the las command which it was 
iptables -t nat -A POSTROUTING -p tcp -d GermanServer--dport 443-j SNAT --to-source GermanServer

**Comment by asoMansoury** [comment]
I got an ubuntu server in Iran, if you are ok please give me you'r email to
send IP and password for ssh to my server and convert it as a bridge to my
German server.

**Reply by trudyhood** [reply]
I hope you can search and manage it yourself. We haven't done it ourselves and are curious to see the results. I believe VpnHood is already compatible with that type of bridge, but you can count on us if you find any part of VpnHood that needs to be modified for better compatibility with that bridge. I am afraid that I can't assist you more at this time.

**Reply by asoMansoury** [reply]
OK. I did what you said. but it didnt work properly.
1 - At first I add port Port Forwarding on my Iran server to forward 443 traffic to my German Server.
2-  Then I went to My Access Point Server  and clicked on add server. 
3 - After that I chose install VpnHood by username and password and set my Iran's IP on hostname part.
4 - When I went to check where VpnHood server was installed I understood that instead of VpnHood been installed on my German Server via Iran's bridge server, it has been installed on my Iran Server.

At the beginning I have to mention that I'm completely pretty sure about Port Forwarding that I had done.
I have to know that are you sure that just by port forwarding of  443 everything will work?
I think when I choose to install by username and port, there is another port that i should've port forwarding.
Do you know Which port is used by Access Point Server when the user going to install VpnHood via this item.

**Reply by trudyhood** [reply]
You should install VpnHoodServer on your german server, and in VpnHood AccessServer give the german server credential to install VpnHoodServer on it. Then go to Server and configure your server like this:
![image](https://user-images.githubusercontent.com/72891124/222557360-a14d6d78-36e3-4b5e-8a26-2a48fba2b9c4.png)

By this, the VpnHoodServer will start on your German IP and listen, but the IP in your tokens will be in your Iran server. Please note that the Listen must be false on the second IP because the VpnHoodServer on your german server can not listen to it. The client will try to connect to your Iran Server because the token points to that IP; if your bridge works, it will be forwarded to the VpnHoodServer on your german server.

Let me know the result.

**Reply by asoMansoury** [reply]
About the above picture you uploaded, can you tell me where is it's menu?
I can't find it in my panel. I always use console.vpnhood.com for managing my servers.

**Reply by asoMansoury** [reply]
I configured Iran's server but I faced this error.
![1](https://user-images.githubusercontent.com/14541559/222648481-35fb04c4-71c2-4410-ba27-655a3207038b.png)


No TcpEndPoint has been configured!

**Reply by trudyhood** [reply]
"Listen" in false, so the server does not listen to anything

**Reply by asoMansoury** [reply]
Now I got this error.
![3](https://user-images.githubusercontent.com/14541559/222668603-09d09fad-3326-4046-b48d-73a102d633a0.png)

Cannot assign requested address.

**Reply by trudyhood** [reply]
I think you install the VpnHood server on your Iran server. You need to install it on your german server. Please double-check it.

**Reply by asoMansoury** [reply]
Yeah you are right. 
I installed VpnHood on German Server, I think when I clicked on Add in Server Panel to add Iran Server Install on Linux by UserName & Password option and it installed VpnHood on my Iran Server automatically!
Which option should have i picked for not installing VpnHood automatically in such case?

what is command to uninstall VpnHood completely on my Iran Server(Ubuntu Server)

**Reply by trudyhood** [reply]
try this:
https://github.com/vpnhood/VpnHood/wiki/How-to-uninstall-VpnHood-Server-from-Linux

**Reply by asoMansoury** [reply]
Hi again
I configured everything based on your guide. Dont install VpnHood Server on Iran, installed VpnHood Server on German on 443 port.
I got a token which is included Iran's IP. When trying to connect, it connects but when i'm going to browse on the internet nothing doesn't open, like the system does not have access to the internet, meanwhile the client have access telegram and instagram, browser doesn't open anything!
do you have any suggestion?

**Reply by trudyhood** [reply]
Do Telegram and Instagram work?
Do you use TCP Mode?
Can you send me an AccessKey to trudy@vpnhood.com?

**Reply by asoMan1369** [reply]
Hi
I fixed it. In such cases clients must choose No Udp in protocol part. by chosing No Udp on client apps, browsers can use VPN perfectly.
Right now I can say your VPN server is the best VPN server I have worked.
Great admin panel, Simplistic implement.
Only thing that I feel is necessary is Max Traffic which doesn't solve yet. By providing this feature VpnHood will be the most and perfect VPN server.

**Reply by asoMan1369** [reply]
If you need a token for checking let me know to send your email

**Comment by trudyhood** [comment]
@asoMan1369 

That's fantastic! Congratulations on your accomplishment.  You can also do the same with that bridge with another VPN than VpnHood too. If you'd like to use UDP mode, it's possible to bridge UDP as well. However, we'll need to modify VpnHood to use one UDP port, as it currently allocates one per user. 

Also, I would really appreciate it if you could create a brief document with step-by-step instructions on the bridge tool and how to use it so we can upload it to our wiki and make it available to others.

Thank you, and I look forward to hearing from you.

**Reply by asoMan1369** [reply]
I really appreciate your VPN Server. By having this feature I can say it is a complete package among all the VpnServers. The reason that it just worked on TCP was because of my PortForwarding that I set on TcpPort. 
About creating  a document, Sure why not. It's my honor to do a tiny help on your great project. I will send you the documentation as soon as possible.

**Reply by zoghal** [reply]
dear @trudyhood 
First of all, I would like to thank you for providing this solution and also for your useful explanations and guidance. I also tested bridge TCP. The result was very good and enjoyable. 🥰

Now I want to ask you a question: Is it possible to implement other strategies? I showed the strategies in the image below:
![vpnhood](https://user-images.githubusercontent.com/174688/223914201-83dbfc67-7679-4da6-b1f8-e612dc5df256.png)

**Reply by asoMan1369** [reply]
> @asoMan1369
> 
> That's fantastic! Congratulations on your accomplishment. You can also do the same with that bridge with another VPN than VpnHood too. If you'd like to use UDP mode, it's possible to bridge UDP as well. However, we'll need to modify VpnHood to use one UDP port, as it currently allocates one per user.
> 
> Also, I would really appreciate it if you could create a brief document with step-by-step instructions on the bridge tool and how to use it so we can upload it to our wiki and make it available to others.
> 
> Thank you, and I look forward to hearing from you.

can I bridge my Iran's server on UDP port?
Actually I need it because while I use UDP port I give maximum net speed from VpnHood.
I remember you have said that you have plan for modifying VpnHood to use One UDP port.
do you have any plan for implementing this feature?

**Reply by zoghal** [reply]
> can I bridge my Iran's server on UDP port? Actually I need it because while I use UDP port I give maximum net speed from VpnHood. I remember you have said that you have plan for modifying VpnHood to use One UDP port. do you have any plan for implementing this feature?

In general, yes, it is possible! But one problem with VPNHood is that the UDP ports are chosen randomly.
If we have the option to define a specific range of UDP ports for VPNHood ! I think this problem can also be solved.

**Reply by asoMansoury** [reply]
> > can I bridge my Iran's server on UDP port? Actually I need it because while I use UDP port I give maximum net speed from VpnHood. I remember you have said that you have plan for modifying VpnHood to use One UDP port. do you have any plan for implementing this feature?
> 
> In general, yes, it is possible! But one problem with VPNHood is that the UDP ports are chosen randomly. If we have the option to define a specific range of UDP ports for VPNHood ! I think this problem can also be solved.

Do you have any plan to add this important feature?
By having this feature we can get the maximum speed from server.

**Reply by trudyhood** [reply]
yes. We have plan to use single UDP port and it is important to us, however I don't know how long does it take.

**Comment by trudyhood** [comment]
Dear @zoghal 
In theory, you can do that; ports are just numbers and do not have any other meaning; for example, you can run HTTP on port 443 or even run HTTPS over port 80. So you can do that and it must work.
Also, I don't know about Linux iptable and have not worked with it yet. Perhaps @asoMan1369 can help you with that.

**Reply by asoMansoury** [reply]
Hello.
About Strategy2 , it seems everything is ok because in this strategy, the bridge server listens to TCP port 8080 and forwards it to the destination server in the USA on 443 port, and this is completely OK.
About the third strategy, I do not think it would be ok, because they used nginx for portwarding, and as you know nginx just works on HTTP and HTTPS protocol. it is so important to work just on TCP protocol not any other till the VPN hood developers provide it on UDP protocol.
As you know VpnHood Server is not a proxy server and I think this is the most powerful point of this server. I think if they work on both TCP/UDP protocol , that would be so nice.
thanks alot.

**Reply by trudyhood** [reply]
I am curious. Who is using nginx?

**Reply by asoMansoury** [reply]
On the above picture in the third scenario, he said wants to do port forwarding by Nginx by listening on 443 HTTP protocol. I think it's wrong, because till i know and looked at your code, you are just working on TCP and UDP, am I right?

**Reply by trudyhood** [reply]
@zoghal
Sorry, I didn't notice. You can not use nginx to forward the port (at least yet), because VpnHoodServer is not compatible with Web Proxies. We have some ideas to support Web Servers like CloudFlares or perhaps nginx but it is not on our table yet.
